File: apps/android/app/src/main/kotlin/com/loonext/android/features/compose/WrapUpDictation.kt
package com.loonext.android.features.compose

import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import android.os.SystemClock
import com.loonext.android.core.auth.await
import com.loonext.android.core.net.ApiClient
import com.loonext.android.core.net.ApiErrorCode
import com.loonext.android.core.net.ApiException
import java.io.File
import java.io.IOException
import java.util.Locale
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * #507 Phase 1 — the wrap-up a crew member speaks after hanging up.
 *
 * They hold a button, say "quoted him $2,400 for the tank, parts Thursday, he's
 * confirming with his wife", and get those words back as text to check and post
 * as an internal note.
 *
 * # Whose voice, and why the answer is the whole feature
 *
 * The member's own, into their own handset, about a call that has ENDED. Never
 * the call, never the customer. D117 is why that line is load-bearing rather
 * than a nicety: every interception statute attaches to the moment the other
 * party's voice is ACQUIRED, so a live-call version is a different feature with
 * a consent architecture this one does not need and does not have. Nothing in
 * this file — and no string it prints — may suggest otherwise.
 *
 * # The audio does not survive the upload
 *
 * That is the promise the feature is sold on, so it is enforced in the one
 * place the client controls: [WrapUpRecorder.finish] reads the cache file into
 * memory and deletes it before it returns, and [WrapUpRecorder.discard] deletes
 * it on every path that does not upload. The bytes then live only as a local in
 * the calling coroutine. The server side of the same promise is in
 * `apps/api/src/ai/call-wrapup.ts`: the audio never reaches R2 and no id exists
 * that could fetch it back.
 *
 * # It is a shortcut, never a precondition
 *
 * Every failure — no permission, no microphone, the toggle off, the monthly cap,
 * a model that said nothing usable — leaves the member exactly where they were:
 * the note composer, with a keyboard. That is why [wrapUpDictationMessage]
 * names what happened and every sentence ends somewhere the member can still
 * act.
 */
object WrapUpDictation {

    /**
     * Longest dictation we record, in seconds. Mirrors CALL_WRAPUP_MAX_SECONDS
     * in `apps/api/src/ai/call-wrapup.ts`, which is the gate that counts —
     * this one exists so a phone left in a pocket is never uploaded at all.
     */
    const val MAX_SECONDS = 120

    /**
     * Largest upload we attempt, in bytes. Mirrors CALL_WRAPUP_MAX_BYTES. The
     * encoder settings below put two minutes at a few hundred KB, so this only
     * ever catches something already wrong.
     */
    const val MAX_BYTES = 8 * 1024 * 1024

    /**
     * Below this the hold was a mis-tap, not a wrap-up.
     *
     * Enforced here rather than left to the server because the server's version
     * of this check happens after the request: a stray brush of the button must
     * not cost a round trip, and it must never cost an audio minute.
     */
    const val MIN_SECONDS = 1

    /**
     * Whole seconds of audio for an elapsed hold.
     *
     * Floored, because a partial second is not a second of speech, and clamped
     * to [MAX_SECONDS] because MediaRecorder's own max-duration stop means the
     * file cannot be longer than that however long the finger stayed down. Both
     * directions are the honest number rather than a flattering one — `seconds`
     * is the client's CLAIM about the audio and the server checks the bytes
     * separately.
     */
    fun elapsedSeconds(elapsedMs: Long): Int {
        if (elapsedMs <= 0L) return 0
        val whole = (elapsedMs / 1000L).toInt()
        return if (whole > MAX_SECONDS) MAX_SECONDS else whole
    }

    /**
     * Is this recording worth spending a request on? The client twin of
     * `shouldTranscribeWrapUp` — same two gates, same reasoning: a recording of
     * nothing, or one that ran away, must not spend.
     */
    fun worthSending(seconds: Int, bytes: Int): Boolean =
        bytes > 0 && bytes <= MAX_BYTES && seconds >= MIN_SECONDS && seconds <= MAX_SECONDS

    /** The running counter shown while the button is held: "0:07", "1:53". */
    fun elapsedLabel(seconds: Int): String =
        String.format(Locale.US, "%d:%02d", seconds / 60, seconds % 60)
}

/**
 * POST /v1/conversations/:id/wrap-up-transcript — what came back.
 *
 * A 200 either carries the words or says why there are none; both are normal
 * answers, so the refusal is DATA rather than an error. See
 * [wrapUpDictationMessage] for the sentence each reason earns.
 */
@Serializable
data class WrapUpTranscript(
    /** The member's words, verbatim. Null means nothing usable came back. */
    val text: String? = null,
    /** Why there are no words; absent on success. */
    val reason: String? = null,
)

/**
 * Plain-language copy for a wrap-up that produced no words.
 *
 * One blanket "something went wrong" would hide a workspace toggle and an
 * exhausted monthly budget behind what reads as a glitch, so each reason says
 * what happened and where the member goes next — and every one of them ends at
 * the keyboard they already have. Mirrors [replyDraftMessage]'s shape, and the
 * reasons themselves are `AiRunFailure` in `apps/api/src/ai/run.ts` plus the
 * route's own `too_long` and `unusable_output`.
 */
fun wrapUpDictationMessage(reason: String?): String = when (reason) {
    "too_long" -> "That was longer than two minutes. Say the short version, or type the note."
    "disabled" -> "Dictated wrap-ups are turned off for this workspace. Settings, AI turns them back on."
    "over_cap" -> "This month's dictation is used up. It starts again next month — type the note for now."
    "model_error", "unavailable" -> "Couldn't write that down just now. Try again, or type the note."
    "unusable_output" -> "Couldn't make out any words. Try again somewhere quieter, or type the note."
    else -> "Couldn't write that down. Type the note instead."
}

/**
 * The one multipart door #507 needs.
 *
 * [ApiClient] speaks JSON bodies and keeps its base URL private, so this
 * borrows its OkHttp client, its fresh-session refresh and its SPEC §7 envelope
 * decoding — exactly as [NoteFileUploader] does next door, and for the same
 * reason.
 *
 * Throws [ApiException] on a non-2xx like every other write in the app, so a
 * signed-out session or a refused capability keeps the server's own sentence
 * instead of being flattened into a reason it does not have. A 200 that carries
 * a `reason` is a refusal the member should read, not a failure, and comes back
 * as data.
 */
class WrapUpTranscriber(private val api: ApiClient, private val baseUrl: String) {

    suspend fun transcribe(
        companyId: String,
        conversationId: String,
        audio: ByteArray,
        seconds: Int,
    ): WrapUpTranscript {
        val session = api.freshSession() ?: throw ApiException(
            ApiErrorCode.UNAUTHORIZED,
            "You're signed out.",
            401,
        )
        val body = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart("seconds", seconds.toString())
            .addFormDataPart(
                "audio",
                WRAP_UP_FILE_NAME,
                audio.toRequestBody(WRAP_UP_CONTENT_TYPE.toMediaTypeOrNull()),
            )
            .build()
        val request = Request.Builder()
            .url("$baseUrl/v1/conversations/$conversationId/wrap-up-transcript")
            .header("Authorization", "Bearer ${session.accessToken}")
            .header("X-Company-Id", companyId)
            .post(body)
            .build()
        val response = try {
            api.http.newCall(request).await()
        } catch (_: IOException) {
            throw ApiException(
                ApiErrorCode.NETWORK,
                "Can't reach Loonext. Check your connection.",
                0,
            )
        }
        val (status, text) = response.use { it.code to it.body?.string().orEmpty() }
        val payload = ApiClient.RawResponse(status, text).expectSuccess(api.json)
        return api.json.decodeFromString(payload)
    }
}

/** AAC in an MPEG-4 container: what MediaRecorder writes below. */
private const val WRAP_UP_CONTENT_TYPE = "audio/mp4"
private const val WRAP_UP_FILE_NAME = "wrap-up.m4a"

/** One finished dictation: the bytes to upload and how long it ran. */
class WrapUpRecording(val audio: ByteArray, val seconds: Int)

/**
 * How a hold ended.
 *
 * Three outcomes rather than a nullable recording, because two of them read
 * identically to the caller and want opposite answers. A brush of the button is
 * a non-event and must be silent — a sentence after every mis-tap is noise. A
 * recorder that died mid-hold, because a call seized the microphone or the
 * permission was revoked from the settings shade, is somebody who watched a
 * counter tick for thirty seconds and then got nothing, and telling them
 * nothing is the worst answer available.
 */
sealed interface WrapUpFinish {
    /** Words worth sending. */
    class Ready(val recording: WrapUpRecording) : WrapUpFinish
    /** Too short or too small to be a wrap-up. Say nothing. */
    data object TooShort : WrapUpFinish
    /** It was running and it broke. Say so. */
    data object Failed : WrapUpFinish
}

/**
 * Records one wrap-up to a cache file and hands back its bytes.
 *
 * NOT thread-safe and not reusable across concurrent holds — one composer, one
 * finger, one recording at a time. Every method is safe to call in the wrong
 * order (a double release, a discard with nothing running); the button is a
 * physical gesture and the states race by nature, so the recorder absorbs that
 * rather than asking the UI to be perfect about it.
 */
class WrapUpRecorder(private val context: Context) {

    private var recorder: MediaRecorder? = null
    private var file: File? = null
    private var startedAtMs = 0L

    /** True between a successful [start] and the [finish]/[discard] that ends it. */
    val isRecording: Boolean get() = recorder != null

    /**
     * Begin recording. Returns false when the platform refused — a microphone
     * held by a call, an OEM that rejects the encoder, a permission revoked
     * between the check and here. Nothing is left running on a false.
     */
    fun start(): Boolean {
        if (recorder != null) return true
        val target = try {
            // cacheDir, deliberately: this file must be the most disposable
            // thing on the device. It is deleted below on every path, and if
            // the process dies mid-hold the system is free to reclaim it —
            // which is the correct outcome for audio nobody promised to keep.
            File.createTempFile("wrap-up-", ".m4a", context.cacheDir)
        } catch (_: IOException) {
            return false
        }
        // PUBLISHED BEFORE THE ARMING BELOW, and that ordering is the whole
        // point. `prepare()` and `start()` take a couple of hundred
        // milliseconds, they run on an IO thread, and `discard()` runs on the
        // main thread from the composition's onDispose. Assigning this field
        // last meant a Back press inside that window found both fields null,
        // deleted nothing, and then watched start() finish and adopt a
        // recorder nobody would ever stop — leaving the audio on the device,
        // which is the one thing this feature promises cannot happen.
        file = target
        val next = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            MediaRecorder(context)
        } else {
            @Suppress("DEPRECATION")
            MediaRecorder()
        }
        return try {
            next.setAudioSource(MediaRecorder.AudioSource.MIC)
            next.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            next.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            // Speech into a handset held at the mouth, sized for a transcript
            // rather than for listening: mono, 16 kHz, 32 kbps puts two minutes
            // at roughly 500 KB. Nobody plays this back — it exists for the
            // length of one request — so bitrate spent above intelligibility is
            // upload time on a job site's signal and nothing else.
            next.setAudioChannels(1)
            next.setAudioSamplingRate(16_000)
            next.setAudioEncodingBitRate(32_000)
            // The platform's own stop, under the UI's timer rather than instead
            // of it: a composition that misses its tick must not leave a
            // microphone open, and neither cap may be the only one.
            next.setMaxDuration(WrapUpDictation.MAX_SECONDS * 1000)
            next.setMaxFileSize(WrapUpDictation.MAX_BYTES.toLong())
            next.setOutputFile(target.absolutePath)
            next.prepare()
            next.start()
            // Did a discard() land while we were arming? It clears `file`, so
            // this is how the losing side of that race finds out. Honour it
            // rather than adopting a recorder the composition has already
            // stopped tracking.
            if (file !== target) {
                runCatching { next.release() }
                target.delete()
                return false
            }
            recorder = next
            startedAtMs = SystemClock.elapsedRealtime()
            true
        } catch (_: Exception) {
            // IllegalStateException, IOException, RuntimeException from start()
            // on a busy mic — all one answer to the caller, which then says so
            // and leaves the keyboard where it was.
            runCatching { next.release() }
            target.delete()
            recorder = null
            file = null
            false
        }
    }

    /** Whole seconds recorded so far, for the counter under the member's thumb. */
    fun elapsedSeconds(): Int =
        if (recorder == null) 0
        else WrapUpDictation.elapsedSeconds(SystemClock.elapsedRealtime() - startedAtMs)

    /**
     * Stop and read the audio out, deleting the file before returning.
     *
     * Null when there is nothing worth uploading: no recording running, a stop
     * the encoder refused (too short to have written a valid header), or a hold
     * that does not clear [WrapUpDictation.worthSending]. The file is gone
     * either way — that is the point of doing the read here rather than handing
     * a path to the caller.
     */
    fun finish(): WrapUpFinish {
        val active = recorder
        if (active == null) {
            // Nothing running, but a file from a start that never reached a
            // stop might still be sitting in the cache. Clear it.
            discard()
            return WrapUpFinish.TooShort
        }
        val target = file
        val seconds = elapsedSeconds()
        val stopped = runCatching { active.stop() }.isSuccess
        runCatching { active.release() }
        recorder = null
        file = null
        val audio = if (stopped && target != null) {
            runCatching { target.readBytes() }.getOrNull()
        } else {
            null
        }
        target?.delete()
        // It WAS running and we could not get the bytes out: a call seized the
        // microphone, the permission was revoked mid-hold, the encoder wrote no
        // usable header. Distinct from a mis-tap, because somebody was waiting.
        if (audio == null) return WrapUpFinish.Failed
        return if (WrapUpDictation.worthSending(seconds, audio.size)) {
            WrapUpFinish.Ready(WrapUpRecording(audio, seconds))
        } else {
            WrapUpFinish.TooShort
        }
    }

    /**
     * Abandon whatever is running and delete the file. Safe to call always,
     * from any thread.
     *
     * Deliberately does NOT call `stop()`. `stop()` finalizes the MPEG-4 moov
     * atom synchronously — hundreds of milliseconds of disk and mediaserver
     * IPC — and this runs on the main thread from the composition's onDispose,
     * which is the navigation frame. `release()` alone tears the recorder down,
     * and a container we are about to delete does not need to be valid.
     */
    fun discard() {
        recorder?.let { active -> runCatching { active.release() } }
        recorder = null
        file?.delete()
        file = null
    }
}

File: apps/ios/Loonext/Features/Thread/WrapUpDictation.swift
import AVFoundation
import Foundation

/// #507 Phase 1 — the wrap-up a crew member SPEAKS after hanging up.
///
/// They have just put the phone down. They hold the mic in the note box, say
/// "quoted him $2,400 for the tank, parts Thursday, he's confirming with his
/// wife", and get those words back as text to check and post as an internal
/// note.
///
/// # Whose voice, and why the whole design is that one line
///
/// D117. This is the MEMBER's own voice, about a call that has ENDED — never
/// the call, and never the customer. Producing a summary of a live two-party
/// call means acquiring the other party's voice, and Canada's Criminal Code
/// s.183 defines "intercept" to include acquiring "the substance, meaning or
/// purport" — which is what a transcript IS. So the live version needs a whole
/// consent architecture (#509) and is not what this is.
///
/// Everything here holds that line mechanically rather than by intention:
///   - Recording only ever starts from a deliberate press-and-hold, and stops
///     the moment the finger lifts. Nothing records in the background.
///   - `WrapUpRecorder.start` REFUSES while a call still holds the line, so
///     the one arrangement that could pick up a customer through the earpiece
///     cannot be entered by accident.
///   - Every string the member reads says the same thing in plain words.
///
/// # The recording does not survive the upload
///
/// `AVAudioRecorder` needs a file, so one is written to the OS temporary
/// directory. `finish()` reads it into memory and DELETES it in the same call —
/// before the upload begins, not after — so a failed or abandoned upload cannot
/// leave audio on the device either. Nothing is ever copied anywhere else and
/// no id exists that could fetch it back; the server drops the bytes the same
/// way (routes/conversations.ts).
///
/// # Best effort, always
///
/// Every failure leaves the member exactly where they were: the note composer,
/// with a keyboard. Dictation is a shortcut, never a precondition, so each
/// reason gets its own plain sentence rather than one blanket shrug — the same
/// argument `replyDraftMessage` documents.

/// The `POST /v1/conversations/:id/wrap-up-transcript` body: the words, or a
/// reason there are none. Both fields optional so either shape decodes and
/// neither ever throws.
struct WrapUpTranscript: Codable, Sendable {
    let text: String?
    /// `too_long`, `disabled`, `over_cap`, `model_error`, `unavailable`,
    /// `unusable_output`. Absent on success. See `wrapUpFailureMessage`.
    var reason: String?
}

/// What the composer got back: words to review, or a sentence to show.
///
/// The failure arm carries a RESOLVED sentence rather than a reason code
/// because two different things produce one: the server's `reason`, and a
/// thrown `ApiError` whose own message (a refused number level, an expired
/// session) is better than anything a code could map to. Resolving both at the
/// one place that holds both keeps the composer down to "show it or use it".
enum WrapUpOutcome: Sendable {
    case text(String)
    case failed(String)
}

/// Plain-language copy for an empty result, one sentence per reason.
///
/// Every one of them ends somewhere the member can still act, because the note
/// composer is right there and typing always works. Mirrors the shape of
/// `replyDraftMessage`.
///
/// #228: the sentences live in `ThreadStrings`, and `locale` defaults to nil so
/// the English is what a caller with no reader in hand still gets. The keys are
/// this app's own (`thread.wrapUpFail…`) rather than Android's flat
/// `thread.wrapUpTooLong` family, because the two clients say these in
/// different words and one key holding two sentences is exactly the drift a
/// shared catalogue exists to stop.
func wrapUpFailureMessage(_ reason: String?, locale: String? = nil) -> String {
    switch reason {
    case "too_long":
        return AppStrings.translate(locale, "thread.wrapUpFailTooLong")
    case "disabled":
        return AppStrings.translate(locale, "thread.wrapUpFailDisabled")
    // #581: billing, not breakage — so it must not say "try again", which is
    // not what fixes it. Same sentence everywhere Lou refuses for this reason,
    // which is why this one DOES share Android's key.
    case "subscription_inactive":
        return AppStrings.translate(locale, "thread.louPausedForBilling")
    case "over_cap":
        return AppStrings.translate(locale, "thread.wrapUpFailOverCap")
    case "model_error", "unavailable":
        return AppStrings.translate(locale, "thread.wrapUpFailUnreachable")
    case "unusable_output":
        return AppStrings.translate(locale, "thread.wrapUpFailUnusable")
    default:
        return AppStrings.translate(locale, "thread.wrapUpFailDefault")
    }
}

/// The client half of the server's gates. Mirrors CALL_WRAPUP_MAX_SECONDS /
/// CALL_WRAPUP_MAX_BYTES in apps/api/src/ai/call-wrapup.ts.
///
/// The server's copies are the ones that COUNT — these exist so a phone left in
/// a pocket is stopped on the device rather than after paying to upload it.
enum WrapUpLimits {
    /// Stop recording here. A wrap-up is a sentence or three; past two minutes
    /// is a stuck finger.
    static let maxSeconds = 120
    /// Refuse to upload anything larger. The server refuses it too.
    static let maxBytes = 8 * 1024 * 1024
    /// Below this a press was a mis-tap, not a dictation. Discarded on the
    /// device: the server would only bill us to tell us the same thing.
    static let minSeconds = 1.0
}

/// What the composer's mic control is doing. Lives here beside the recorder so
/// the two names are read together, but the composer OWNS the value — see the
/// note on `WrapUpRecorder`.
enum WrapUpPhase: Equatable, Sendable {
    case idle
    case recording
    case transcribing
}

/// Where the microphone permission stands right now, without asking for it.
/// A small local enum rather than AVFoundation's, so the composer never has to
/// import AVFoundation to draw a button.
enum WrapUpMicPermission: Equatable, Sendable {
    case granted
    case denied
    case unasked
}

/// Everything the composer needs to offer dictation, as ONE parameter.
///
/// Grouped for the reason `DuplicateReplyContext` documents just above it in
/// Composer.swift: that view's call site has run the Swift type checker out of
/// budget before, and two more closures is exactly the shape that does it.
struct WrapUpDictationContext {
    /// True while a call still holds a line. Checked at PRESS time rather than
    /// read once into a disabled state, because the answer changes underneath
    /// the composer — a member can dismiss the in-call screen and land back on
    /// this thread with the call still up.
    let callInProgress: @MainActor () -> Bool
    /// Upload the dictation and hand back words or a sentence. Never throws:
    /// a failure here must cost a shortcut, never the note.
    let transcribe: @MainActor (Data, Int) async -> WrapUpOutcome
}

/// Why a press could not start recording, and the sentence it shows.
///
/// Top level rather than nested in the recorder so the copy is readable without
/// touching AVFoundation — every one of these is a claim about what this
/// feature does, and D117 makes those the strings most worth pinning.
enum WrapUpStartRefusal: Equatable, Sendable {
    case callInProgress
    case micDenied
    case micJustGranted
    case couldNotStart

    /// #228 — the catalogue key, so the reason and the sentence stay apart.
    var messageKey: String {
        switch self {
        case .callInProgress:
            // The one refusal that is about D117 rather than about a device.
            // Says what to do next, and says it without ever suggesting the
            // call itself could be written down.
            return "thread.wrapUpRefusalCallInProgress"
        case .micDenied:
            // iOS's own permission path, which is why this is not Android's
            // `thread.micDeniedWrapUp`: a member sent to
            // "Settings › Apps › Loonext › Permissions" on an iPhone is being
            // sent somewhere that does not exist.
            return "thread.wrapUpRefusalMicDenied"
        case .micJustGranted:
            // The permission sheet ate the press: iOS asks, the member
            // answers, and by then whatever they said is gone. Saying so beats
            // an empty note and beats silence.
            return "thread.wrapUpRefusalMicJustGranted"
        case .couldNotStart:
            return "thread.wrapUpRefusalCouldNotStart"
        }
    }

    /// The sentence, in the reader's language.
    ///
    /// A method with its OWN name rather than an overload of `message`: that
    /// property is read through a key path (`refusals.map(\.message)`), and a
    /// property and a method sharing a base name is the kind of thing this
    /// project cannot compile locally to find out about.
    func localizedMessage(_ locale: String?) -> String {
        AppStrings.translate(locale, messageKey)
    }

    /// The English, for callers with no reader in hand.
    var message: String { localizedMessage(nil) }
}

/// Records one wrap-up to a temporary file and hands back its bytes.
///
/// A plain @MainActor class rather than an `@Observable` one: the composer owns
/// what it DRAWS (the phase, the seconds) as its own state, and this owns only
/// what SwiftUI cannot — the live `AVAudioRecorder` and the file underneath it.
/// Two owners of one truth is how a button ends up drawn as recording after the
/// recorder has stopped.
@MainActor
final class WrapUpRecorder {
    private var recorder: AVAudioRecorder?
    private var fileURL: URL?
    /// Whether WE activated the shared audio session. Tracked rather than
    /// assumed, because `discard()` is a safe-to-call-anytime cleanup: without
    /// this flag, tidying up after a dictation that finished ten minutes ago
    /// would deactivate the session a CALL is currently holding.
    private var sessionActive = false

    /// True while audio is being captured.
    var isRecording: Bool { recorder != nil }

    /// Where the microphone permission stands, without asking for it. Read on
    /// every press so a member who allowed it in Settings mid-session is not
    /// stuck behind a cached answer.
    ///
    /// A `default` arm rather than an exhaustive one, the same rule the wire
    /// models follow: a case Apple adds later must degrade to "ask", never
    /// fail to compile and never silently read as granted.
    var micPermission: WrapUpMicPermission {
        switch AVAudioApplication.shared.recordPermission {
        case .granted: return .granted
        case .denied: return .denied
        default: return .unasked
        }
    }

    /// Ask for the microphone, at the point of use. True when allowed.
    ///
    /// The grant never records THIS press: iOS puts its sheet up while the
    /// member is already talking, so whatever they said is gone. The caller
    /// says `micJustGranted` instead of pretending — an empty note would be
    /// worse than a sentence asking them to say it again.
    func requestMic() async -> Bool {
        await AVAudioApplication.requestRecordPermission()
    }

    /// Begin capturing. Returns nil on success, or the sentence to show.
    ///
    /// `callInProgress` is passed in rather than inspected here so the check
    /// uses the softphone's own state (the only honest source) instead of
    /// guessing from the audio session, whose category can outlive a call.
    func start(callInProgress: Bool) -> WrapUpStartRefusal? {
        if isRecording { return nil }
        if callInProgress { return .callInProgress }

        // `.record`, never `.playAndRecord`: this app treats a `.playAndRecord`
        // session as "a call owns the audio" (see AttachmentMedia's player), so
        // borrowing that category for a dictation would make an attachment play
        // out of the earpiece afterwards.
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.record, mode: .spokenAudio)
            try session.setActive(true)
            sessionActive = true
        } catch {
            return .couldNotStart
        }

        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("wrap-up-\(UUID().uuidString).m4a")
        // Mono 16 kHz AAC: what speech recognition wants and nothing more, so
        // a two-minute ceiling lands far inside the server's 8 MB one.
        let settings: [String: Any] = [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 16000.0,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.medium.rawValue,
        ]
        let made: AVAudioRecorder
        do {
            made = try AVAudioRecorder(url: url, settings: settings)
        } catch {
            deactivate()
            return .couldNotStart
        }
        guard made.record() else {
            deactivate()
            try? FileManager.default.removeItem(at: url)
            return .couldNotStart
        }
        recorder = made
        fileURL = url
        return nil
    }

    /// Stop and hand back the bytes, DELETING the file in the same call.
    ///
    /// Nil means there is nothing worth uploading — a mis-tap, an unreadable
    /// file, or something past the size the server would refuse anyway. The
    /// file is gone in every one of those cases too.
    /// Delete any wrap-up audio a previous run left behind.
    ///
    /// Every in-process path deletes its own file, but none of them runs if the
    /// app is killed mid-hold — a crash, a swipe-up, the system reclaiming
    /// memory behind a phone call. The temporary directory is not guaranteed to
    /// be emptied on any schedule the user can see, so without this a recording
    /// can sit on the device indefinitely, which is the one thing this feature
    /// promises cannot happen.
    ///
    /// Named by prefix so it can only ever remove this feature's own files.
    static func sweepOrphans() {
        let fm = FileManager.default
        let dir = fm.temporaryDirectory
        guard let entries = try? fm.contentsOfDirectory(
            at: dir,
            includingPropertiesForKeys: nil
        ) else { return }
        for url in entries where url.lastPathComponent.hasPrefix("wrap-up-") {
            try? fm.removeItem(at: url)
        }
    }

    func finish() -> (audio: Data, seconds: Int)? {
        guard let made = recorder else {
            discard()
            return nil
        }
        // Read the duration BEFORE stopping; `currentTime` is 0 afterwards.
        let duration = made.currentTime
        made.stop()
        recorder = nil

        var bytes: Data?
        if let fileURL {
            bytes = try? Data(contentsOf: fileURL)
        }
        // The recording does not survive this method, whatever happened above.
        discard()

        guard duration >= WrapUpLimits.minSeconds, let audio = bytes else { return nil }
        guard !audio.isEmpty, audio.count <= WrapUpLimits.maxBytes else { return nil }
        // Clamped to the cap, not merely measured against it. The auto-stop
        // fires AT the ceiling and this runs a moment later, so an unclamped
        // duration reads 121 for a recording we deliberately ended at 120 — and
        // the server refuses it. The one dictation the app stopped on the
        // member's behalf would have been the one guaranteed to fail, after
        // paying to upload it.
        let clamped = min(WrapUpLimits.maxSeconds, max(1, Int(duration.rounded())))
        return (audio: audio, seconds: clamped)
    }

    /// Throw the recording away unread — a cancelled press, or a member who
    /// left the thread mid-sentence. Safe to call when nothing is recording.
    func discard() {
        recorder?.stop()
        recorder = nil
        if let fileURL {
            try? FileManager.default.removeItem(at: fileURL)
        }
        fileURL = nil
        deactivate()
    }

    /// Hand the audio route back, and only if we took it. The category is
    /// deliberately left at `.record`: the attachment player re-sets
    /// `.playback` before it plays, and the softphone sets its own when a call
    /// starts.
    private func deactivate() {
        guard sessionActive else { return }
        sessionActive = false
        try? AVAudioSession.sharedInstance().setActive(
            false,
            options: [.notifyOthersOnDeactivation]
        )
    }
}

/// POST /v1/conversations/:id/wrap-up-transcript — the one door.
///
/// On `MultipartClient` because `ApiClient` speaks only JSON bodies and this is
/// a file upload; the same place `uploadNoteFile` lives. NEVER throws: a
/// wrap-up that fails must leave the composer exactly as it was.
extension MultipartClient {
    func wrapUpTranscript(
        companyId: String,
        conversationId: String,
        audio: Data,
        seconds: Int,
        /// #228 — declared LAST and defaulted, so every existing call site is
        /// unchanged and the sentences below reach the reader in their language.
        locale: String? = nil
    ) async -> WrapUpOutcome {
        let raw: Data
        do {
            raw = try await postFile(
                path: "/v1/conversations/\(conversationId)/wrap-up-transcript",
                companyId: companyId,
                fields: [("seconds", String(seconds))],
                fileField: "audio",
                fileName: "wrap-up.m4a",
                contentType: "audio/mp4",
                bytes: audio
            )
        } catch {
            // The server's own words when it has any — a refused number level
            // or an expired session says far more than "model_error" would.
            return .failed(error.userMessage)
        }
        guard let decoded = try? JSONDecoder().decode(WrapUpTranscript.self, from: raw) else {
            return .failed(wrapUpFailureMessage("unusable_output", locale: locale))
        }
        guard let text = decoded.text, !text.isBlank else {
            return .failed(wrapUpFailureMessage(decoded.reason, locale: locale))
        }
        return .text(text)
    }
}
